sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -633,25 +618,26 @@ trait HiveTypeCoercion {
     import HiveTypeCoercion._
 
     def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
-      case cw: CaseWhenLike if !cw.resolved && cw.childrenResolved && !cw.valueTypesEqual =>
+      case cw: CaseWhenLike if cw.childrenResolved && cw.checkInputDataTypes().hasError =>
         logDebug(s"Input values for null casting ${cw.valueTypes.mkString(",")}")
-        val commonType = cw.valueTypes.reduce { (v1, v2) =>
-          findTightestCommonType(v1, v2).getOrElse(sys.error(
-            s"Types in CASE WHEN must be the same or coercible to a common type: $v1 != $v2"))
-        }
-        val transformedBranches = cw.branches.sliding(2, 2).map {
-          case Seq(when, value) if value.dataType != commonType =>
-            Seq(when, Cast(value, commonType))
-          case Seq(elseVal) if elseVal.dataType != commonType =>
-            Seq(Cast(elseVal, commonType))
-          case s => s
-        }.reduce(_ ++ _)
-        cw match {
-          case _: CaseWhen =>
-            CaseWhen(transformedBranches)
-          case CaseKeyWhen(key, _) =>
-            CaseKeyWhen(key, transformedBranches)
-        }
+        cw.valueTypes.foldLeft[Option[DataType]](Some(NullType))((r, c) => r match {
+          case None => None
+          case Some(d) => findTightestCommonType(d, c)
+        }).map { commonType =>
+          val transformedBranches = cw.branches.sliding(2, 2).map {
+            case Seq(when, value) if value.dataType != commonType =>
+              Seq(when, Cast(value, commonType))
+            case Seq(elseVal) if elseVal.dataType != commonType =>
+              Seq(Cast(elseVal, commonType))
+            case s => s
+          }.reduce(_ ++ _)
+          cw match {
+            case _: CaseWhen =>
+              CaseWhen(transformedBranches)
+            case CaseKeyWhen(key, _) =>
+              CaseKeyWhen(key, transformedBranches)
+          }
+        }.getOrElse(cw)
```

Comment:
Shouldn't we throw an exception here since `cw.checkInputDataTypes().hasError` is true?